### PR TITLE
[CARBONDATA-3898] Support Option 'carbon.enable.mv'

### PR DESCRIPTION
**Pull request by marchpure**
### Why is this PR needed?
When MV enabled, SQL rewrite takes a lot of time, a new option 'carbon.enable.querywithmv' shall be supported, which can turn off SQL Rewrite when the configured value is false
 
 ### What changes were proposed in this PR?
Add option 'carbon.enable.querywithmv', then logicplan won't be changed if configured value is false
    
 ### Does this PR introduce any user interface change?
 - No

 ### Is any new testcase added?
 - Yes

### CODE REVIEW

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1620/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3361/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 1]
```diff
@@ -1762,6 +1762,20 @@ public boolean isDistributedPruningEnabled(String dbName, String tableName) {
...
+   * Check whether the MV is enabled by the user or not.
+   */
+  public boolean isMVEnabled() {
+    String mvEnabled = CarbonProperties.getInstance().getProperty(
+            CarbonCommonConstants.CARBON_ENABLE_QUERYWITHMV);
+    if (mvEnabled == null) {
+      return Boolean.parseBoolean(CarbonCommonConstants.CARBON_ENABLE_QUERYWITHMV_DEFAULT);
+    } else {
+      // return false only if false is set. any other case return true
+      return !mvEnabled.equalsIgnoreCase("false");
```
Comment (kevinjmh) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java:1775]:
```suggestion
      return mvEnabled.equalsIgnoreCase("true");
```

File: integration/spark/src/test/scala/org/apache/carbondata/view/MVTest.scala [hunk 2]
```diff
@@ -63,6 +63,30 @@ class MVTest extends QueryTest with BeforeAndAfterAll {
...
+
+    CarbonProperties.getInstance().addProperty(CarbonCommonConstants.CARBON_ENABLE_QUERYWITHMV,"false")
+    var df = sql("select empname, avg(salary) from source group by empname")
+    assert(!isTableAppearedInPlan(df.queryExecution.optimizedPlan, "mv1"))
+
+    df = sql("select empname, avg(salary) from source group by empname")
+    CarbonProperties.getInstance().removeProperty(CarbonCommonConstants.CARBON_ENABLE_QUERYWITHMV)
+    assert(isTableAppearedInPlan(df.queryExecution.optimizedPlan, "mv1"))
+
+    sql("set carbon.mv.enable = false")
```
Comment (kevinjmh) [inline integration/spark/src/test/scala/org/apache/carbondata/view/MVTest.scala:81]: not consist with the property name added. And I think this one is better than `carbon.enable.querywithmv`

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala [hunk 3]
```diff
@@ -98,14 +98,8 @@ class MVRewriteRule(session: SparkSession) extends Rule[LogicalPlan] {
...
       return logicalPlan
     }
-    val sessionInformation = ThreadLocalSessionInfo.getCarbonSessionInfo
-    if (sessionInformation != null && sessionInformation.getThreadParams != null) {
-      val disableViewRewrite = sessionInformation.getThreadParams.getProperty(
-        CarbonCommonConstants.DISABLE_SQL_REWRITE)
-      if (disableViewRewrite != null &&
-        disableViewRewrite.equalsIgnoreCase("true")) {
-        return logicalPlan
-      }
```
Comment (kevinjmh) [inline integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala:108]: please check this changes again, for the case multiple sessions with different setting

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java (hunk 1 above)
Comment (marchpure) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java]: modified

File: integration/spark/src/test/scala/org/apache/carbondata/view/MVTest.scala (hunk 2 above)
Comment (marchpure) [reply integration/spark/src/test/scala/org/apache/carbondata/view/MVTest.scala]: modified

File: integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala (hunk 3 above)
Comment (marchpure) [reply integration/spark/src/main/scala/org/apache/spark/sql/optimizer/MVRewriteRule.scala:108]: checked, some testcases are added

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3365/

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1624/

**Comment by marchpure** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1627/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3368/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3383/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1642/

File: core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java [hunk 4]
```diff
@@ -501,6 +502,25 @@ private void validateCarbonTaskDistribution() {
...
 
+  private void validateEnableMV() {
+    String isMVEnabled = carbonProperties.getProperty(CarbonCommonConstants.CARBON_ENABLE_MV);
+    if (isMVEnabled == null) {
+      carbonProperties.setProperty(CarbonCommonConstants.CARBON_ENABLE_MV,
+          CarbonCommonConstants.CARBON_ENABLE_MV_DEFAULT);
+      isMVEnabled = carbonProperties.getProperty(CarbonCommonConstants.CARBON_ENABLE_MV);
+    }
+    boolean isValidBooleanValue = CarbonUtil.validateBoolean(isMVEnabled);
+    if (!isValidBooleanValue) {
```
Comment (xubo245) [inline core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java:513]: suggestion:  merge it with line 507,  null also is invalid value.

Comment (xubo245) [reply core/src/main/java/org/apache/carbondata/core/util/CarbonProperties.java:513]: CarbonUtil.validateBoolean  include judge null:

```
  public static boolean validateBoolean(String value) {
    if (null == value) {
      return false;
    } else if (!("false".equalsIgnoreCase(value) || "true".equalsIgnoreCase(value))) {
      return false;
    }
    return true;
  }
```

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 5]
```diff
@@ -2433,12 +2441,6 @@ private CarbonCommonConstants() {
    */
   public static final String INDEX_STATUS = "index_status";
 
-  /**
```
Comment (xubo245) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:2436]: Are there any user already use this configuration?

File: core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java [hunk 6]
```diff
@@ -2433,12 +2441,6 @@ private CarbonCommonConstants() {
...
-   * Materialized view thread context properties
-   */
-  @CarbonProperty
-  public static final String DISABLE_SQL_REWRITE = "disable_sql_rewrite";
```
Comment (xubo245) [inline core/src/main/java/org/apache/carbondata/core/constants/CarbonCommonConstants.java:2440]: Are there any user already use this configuration?

**Comment by CarbonDataQA1** [comment]
Build Failed  with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3386/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1645/

**Comment by marchpure** [comment]
retest this please

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.3.4, Please check CI http://121.244.95.60:12545/job/ApacheCarbonPRBuilder2.3/3388/

**Comment by CarbonDataQA1** [comment]
Build Success with Spark 2.4.5, Please check CI http://121.244.95.60:12545/job/ApacheCarbon_PR_Builder_2.4.5/1647/

**Comment by xubo245** [comment]
LGTM

**Comment by kunal642** [comment]
LGTM

**Pull request closed without merging**